### PR TITLE
rafthttp: report MsgSnap status

### PR DESCRIPTION
**Pull request by yichengq**
for #2381

### CODE REVIEW

**Comment by xiang90** [comment]
have you tested this manually?

**Comment by yichengq** [comment]
Yes. I start node1 and node2 of a 3-member cluster, and boom them to make a snapshot, and add node3 into the cluster:
17:09:22 etcd1 | 2015/03/01 17:09:22 rafthttp: roundtripping error: error roundtripping to http://localhost:7003/raft/stream/message/a8266ecf031671f3: dial tcp 127.0.0.1:7003: connection refused
17:09:22 etcd1 | 2015/03/01 17:09:22 rafthttp: roundtripping error: error roundtripping to http://localhost:7003/raft/stream/a8266ecf031671f3: dial tcp 127.0.0.1:7003: connection refused
17:09:22 etcd2 | 2015/03/01 17:09:22 rafthttp: roundtripping error: error roundtripping to http://localhost:7003/raft/stream/message/6e3bd23ae5f1eae0: dial tcp 127.0.0.1:7003: connection refused
17:09:22 etcd2 | 2015/03/01 17:09:22 rafthttp: roundtripping error: error roundtripping to http://localhost:7003/raft/stream/6e3bd23ae5f1eae0: dial tcp 127.0.0.1:7003: connection refused
17:09:22 etcd2 | 2015/03/01 17:09:22 raft: 6e3bd23ae5f1eae0 [firstindex: 30146, commit: 30347] sent snapshot[index: 30145, term: 11] to 924e2e83e93f2560 [next = 15339, match = 0, wait = 0]
**17:09:22 etcd2 | 2015/03/01 17:09:22 raft: 6e3bd23ae5f1eae0 paused sending replication messages to 924e2e83e93f2560 [next = 15339, match = 0, wait = 0]
17:09:22 etcd2 | 2015/03/01 17:09:22 raft: 6e3bd23ae5f1eae0 snapshot failed, resumed sending replication messages to 924e2e83e93f2560 [next = 15339, match = 0, wait = 0]
17:09:22 etcd2 | 2015/03/01 17:09:22 raft: 6e3bd23ae5f1eae0 [firstindex: 30146, commit: 30347] sent snapshot[index: 30145, term: 11] to 924e2e83e93f2560 [next = 15339, match = 0, wait = 0]
17:09:22 etcd2 | 2015/03/01 17:09:22 raft: 6e3bd23ae5f1eae0 paused sending replication messages to 924e2e83e93f2560 [next = 15339, match = 0, wait = 0]
17:09:22 etcd2 | 2015/03/01 17:09:22 raft: 6e3bd23ae5f1eae0 snapshot failed, resumed sending replication messages to 924e2e83e93f2560 [next = 15339, match = 0, wait = 0]**
17:09:22 etcd1 | 2015/03/01 17:09:22 rafthttp: roundtripping error: error roundtripping to http://localhost:7003/raft/stream/a8266ecf031671f3: dial tcp 127.0.0.1:7003: connection refused
17:09:22 etcd1 | 2015/03/01 17:09:22 rafthttp: roundtripping error: error roundtripping to http://localhost:7003/raft/stream/message/a8266ecf031671f3: dial tcp 127.0.0.1:7003: connection refused
17:09:22 etcd2 | 2015/03/01 17:09:22 rafthttp: roundtripping error: error roundtripping to http://localhost:7003/raft/stream/message/6e3bd23ae5f1eae0: dial tcp 127.0.0.1:7003: connection refused
17:09:22 etcd2 | 2015/03/01 17:09:22 rafthttp: roundtripping error: error roundtripping to http://localhost:7003/raft/stream/6e3bd23ae5f1eae0: dial tcp 127.0.0.1:7003: connection refused
17:09:22 etcd1 | 2015/03/01 17:09:22 rafthttp: roundtripping error: error roundtripping to http://localhost:7003/raft/stream/a8266ecf031671f3: dial tcp 127.0.0.1:7003: connection refused
17:09:22 etcd1 | 2015/03/01 17:09:22 rafthttp: roundtripping error: error roundtripping to http://localhost:7003/raft/stream/message/a8266ecf031671f3: dial tcp 127.0.0.1:7003: connection refused
17:09:22 etcd2 | 2015/03/01 17:09:22 rafthttp: roundtripping error: error roundtripping to http://localhost:7003/raft/stream/message/6e3bd23ae5f1eae0: dial tcp 127.0.0.1:7003: connection refused
17:09:22 etcd2 | 2015/03/01 17:09:22 rafthttp: roundtripping error: error roundtripping to http://localhost:7003/raft/stream/6e3bd23ae5f1eae0: dial tcp 127.0.0.1:7003: connection refused
17:09:22 etcd1 | 2015/03/01 17:09:22 rafthttp: roundtripping error: error roundtripping to http://localhost:7003/raft/stream/a8266ecf031671f3: dial tcp 127.0.0.1:7003: connection refused
17:09:22 etcd1 | 2015/03/01 17:09:22 rafthttp: roundtripping error: error roundtripping to http://localhost:7003/raft/stream/message/a8266ecf031671f3: dial tcp 127.0.0.1:7003: connection refused
17:09:22 etcd2 | 2015/03/01 17:09:22 pipeline: the connection with 924e2e83e93f2560 became active
17:09:22 etcd2 | 2015/03/01 17:09:22 raft: 6e3bd23ae5f1eae0 [firstindex: 30146, commit: 30348] sent snapshot[index: 30145, term: 11] to 924e2e83e93f2560 [next = 15339, match = 0, wait = 0]
**17:09:22 etcd2 | 2015/03/01 17:09:22 raft: 6e3bd23ae5f1eae0 paused sending replication messages to 924e2e83e93f2560 [next = 15339, match = 0, wait = 0]
17:09:22 etcd2 | 2015/03/01 17:09:22 raft: 6e3bd23ae5f1eae0 snapshot succeeded, resumed sending replication messages to 924e2e83e93f2560 [next = 30146, match = 0, wait = 0]**
17:09:23 etcd2 | 2015/03/01 17:09:23 raft: 6e3bd23ae5f1eae0 received msgApp rejection(lastindex: 30145) from 924e2e83e93f2560 for index 30350
17:09:23 etcd2 | 2015/03/01 17:09:23 raft: 6e3bd23ae5f1eae0 decreased progress of 924e2e83e93f2560 to [next = 30146, match = 30145, wait = 0]

File: rafthttp/peer.go [hunk 1]
```diff
@@ -180,6 +180,11 @@ func (p *peer) Stop() {
 
 func (p *peer) pick(m raftpb.Message) (writec chan raftpb.Message, name string, size int) {
 	switch {
+	// Considering MsgSnap may have a big size, e.g., 1G, and blocks
+	// following messages for a long time, only use pipeline to send MsgSnap.
```
Comment (xiang90) [inline rafthttp/peer.go]: only use one of the N pipelines to send...

**Comment by xiang90** [comment]
LGTM.

File: rafthttp/peer.go [hunk 2]
```diff
@@ -180,6 +180,11 @@ func (p *peer) Stop() {
 
 func (p *peer) pick(m raftpb.Message) (writec chan raftpb.Message, name string, size int) {
 	switch {
+	// Considering MsgSnap may have a big size, e.g., 1G, and blocks
```
Comment (xiang90) [inline rafthttp/peer.go]: i think we need to explicitly say it will block a stream for a long time.

File: rafthttp/pipeline.go [hunk 3]
```diff
@@ -102,6 +105,9 @@ func (p *pipeline) handle() {
 			if m.Type == raftpb.MsgApp {
 				p.fs.Fail()
 			}
+			if m.Type == raftpb.MsgSnap {
```
Comment (xiang90) [inline rafthttp/pipeline.go]: isMsgSnap

File: rafthttp/pipeline.go [hunk 4]
```diff
@@ -111,6 +117,9 @@ func (p *pipeline) handle() {
 			if m.Type == raftpb.MsgApp {
 				p.fs.Succ(end.Sub(start))
 			}
+			if m.Type == raftpb.MsgSnap {
```
Comment (xiang90) [inline rafthttp/pipeline.go]: isMsgSnap

**Comment by yichengq** [comment]
all comments addressed.

**Comment by xiang90** [comment]
lgtm

**Pull request merged**